### PR TITLE
Add the dueling network to DQNAgent

### PR DESCRIPTION
**Pull request by raopku**
i try to add the dueling network to dqn. I hope to get your advice

here is how i do it:

- get the second last layer of model passed to the DQNAgent, abandon the last layer

- add a new layer after the second last layer of model with the size (nb_action+1,), which is named as y

1.             # layer y has a shape (nb_action+1,)
2.             # y[:,0] represents V(s;theta)
3.             # y[:,1:] represents A(s,a;theta)

- caculate the Q(s,a;theta) layer as a keras Lambda layer, which is named as outputlayer

1.             # when dueling_type == 'avg'
2.             # Q(s,a;theta) = V(s;theta) + (A(s,a;theta)-Avg_a(A(s,a;theta)))
3.             # when dueling_type == 'max'
4.             # Q(s,a;theta) = V(s;theta) + (A(s,a;theta)-max_a(A(s,a;theta)))
5.             # when dueling_type == 'naive'
6.             # Q(s,a;theta) = V(s;theta) + A(s,a;theta)

- construct a new model

1.            # newmodel = Model(input =model.input, output = outputlayer)

- i write an example of duel dqn for OpenAI CartPole-v0, here is the part of the training logs 

_1.   4260/50000: episode: 125, duration: 1.914s, episode steps: 237, steps per second: 124, episode reward: 237.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.515 [0.000, 1.000], mean observation: 0.258 [-1.459, 2.789], loss: 1.063010, mean_absolute_error: 15.296993, mean_q: 31.297831
2.   4409/50000: episode: 126, duration: 1.223s, episode steps: 149, steps per second: 122, episode reward: 149.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.470 [0.000, 1.000], mean observation: -0.246 [-1.635, 0.717], loss: 1.324580, mean_absolute_error: 16.253447, mean_q: 33.235760
3.   4563/50000: episode: 127, duration: 1.223s, episode steps: 154, steps per second: 126, episode reward: 154.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.461 [0.000, 1.000], mean observation: -0.318 [-2.154, 0.801], loss: 1.413017, mean_absolute_error: 17.135771, mean_q: 35.006638
4.   4811/50000: episode: 128, duration: 1.930s, episode steps: 248, steps per second: 128, episode reward: 248.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.480 [0.000, 1.000], mean observation: -0.160 [-1.783, 0.775], loss: 1.362046, mean_absolute_error: 18.279867, mean_q: 37.348125
5.   5050/50000: episode: 129, duration: 1.885s, episode steps: 239, steps per second: 127, episode reward: 239.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.469 [0.000, 1.000], mean observation: -0.258 [-2.790, 0.984], loss: 1.386958, mean_absolute_error: 19.593142, mean_q: 40.091267
6.   5429/50000: episode: 130, duration: 3.036s, episode steps: 379, steps per second: 125, episode reward: 379.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.512 [0.000, 1.000], mean observation: 0.173 [-1.217, 2.762], loss: 1.914814, mean_absolute_error: 21.176933, mean_q: 43.169743
7.   5614/50000: episode: 131, duration: 1.462s, episode steps: 185, steps per second: 127, episode reward: 185.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.465 [0.000, 1.000], mean observation: -0.336 [-2.584, 0.671], loss: 1.634199, mean_absolute_error: 22.802450, mean_q: 46.470306
8.   5798/50000: episode: 132, duration: 1.465s, episode steps: 184, steps per second: 126, episode reward: 184.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.516 [0.000, 1.000], mean observation: 0.341 [-0.948, 2.276], loss: 1.890901, mean_absolute_error: 23.659718, mean_q: 48.169544
9.   5999/50000: episode: 133, duration: 1.629s, episode steps: 201, steps per second: 123, episode reward: 201.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.517 [0.000, 1.000], mean observation: 0.312 [-0.773, 2.291], loss: 2.066004, mean_absolute_error: 24.598412, mean_q: 50.073372
10.   6175/50000: episode: 134, duration: 1.431s, episode steps: 176, steps per second: 123, episode reward: 176.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.466 [0.000, 1.000], mean observation: -0.308 [-2.174, 0.691], loss: 1.886971, mean_absolute_error: 25.594563, mean_q: 52.077305
11.   6386/50000: episode: 135, duration: 1.747s, episode steps: 211, steps per second: 121, episode reward: 211.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.464 [0.000, 1.000], mean observation: -0.297 [-2.962, 0.804], loss: 1.878460, mean_absolute_error: 26.215256, mean_q: 53.484360
12.   6626/50000: episode: 136, duration: 1.892s, episode steps: 240, steps per second: 127, episode reward: 240.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.471 [0.000, 1.000], mean observation: -0.261 [-2.796, 0.603], loss: 2.331970, mean_absolute_error: 27.323385, mean_q: 55.505970
13.   6877/50000: episode: 137, duration: 2.010s, episode steps: 251, steps per second: 125, episode reward: 251.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.518 [0.000, 1.000], mean observation: 0.247 [-0.890, 2.414], loss: 2.358430, mean_absolute_error: 28.530544, mean_q: 58.037045
14.   7175/50000: episode: 138, duration: 2.377s, episode steps: 298, steps per second: 125, episode reward: 298.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.473 [0.000, 1.000], mean observation: -0.211 [-2.926, 0.878], loss: 2.627178, mean_absolute_error: 29.701906, mean_q: 60.317657
15.   7416/50000: episode: 139, duration: 1.949s, episode steps: 241, steps per second: 124, episode reward: 241.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.523 [0.000, 1.000], mean observation: 0.263 [-0.823, 2.406], loss: 2.652330, mean_absolute_error: 30.846329, mean_q: 62.637028
16.   7578/50000: episode: 140, duration: 1.324s, episode steps: 162, steps per second: 122, episode reward: 162.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.463 [0.000, 1.000], mean observation: -0.369 [-2.767, 0.943], loss: 2.385495, mean_absolute_error: 31.783854, mean_q: 64.615829
17.   7892/50000: episode: 141, duration: 2.503s, episode steps: 314, steps per second: 125, episode reward: 314.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.513 [0.000, 1.000], mean observation: 0.202 [-0.676, 2.407], loss: 2.836606, mean_absolute_error: 32.634499, mean_q: 66.321106
18.   8068/50000: episode: 142, duration: 1.428s, episode steps: 176, steps per second: 123, episode reward: 176.000, mean reward: 1.000 [1.000, 1.000], mean action: 0.460 [0.000, 1.000], mean observation: -0.355 [-2.573, 0.752], loss: 2.669178, mean_absolute_error: 33.854343, mean_q: 68.686623_

### CODE REVIEW

**Comment by matthiasplappert** [comment]
Thanks for this contribution! I've annotated your changes with a few minor concerns and suggestions. Could you please also do the following:

- Please add a reference to the dueling DQN paper to the `README.md` under "References" and also under "What is included?". Just follow the example of the other algorithms that are already listed.
- Please add an integration test under `tests/integration/test_discrete.py`. Essentially you'll want to copy & paste the `test_dqn` method and adjust it so that your dueling DQN implementation is enabled.

Again, thank you for your contribution!

**Comment by matthiasplappert** [comment]
Regarding the problem with CI: let's get everything into a mergable state and I'll try to figure out what exactly is wrong with the theano import. This is clearly unrelated to your changes so don't worry about it too much for now!

**Comment by raopku** [comment]
@matthiasplappert 
Thank you very much for your advices !
Some updates have been added to the pull request. Please check if it is ok.

**Comment by matthiasplappert** [comment]
Great! Can you please also address the remaining comments that I left (I annotated the respective line directly, see above). Thanks!

File: examples/duel_dqn_cartpole.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+# Finally, we configure and compile our agent. You can use every built-in Keras optimizer and
+# even the metrics!
+memory = SequentialMemory(limit=50000, window_length=1)
+policy = BoltzmannQPolicy()
+# enable the dueling network
+# you can specify the dueling_type to one of {'avg','max','naive'}
+dqn = DQNAgent(model=model, nb_actions=nb_actions, memory=memory, nb_steps_warmup=10,
+               enable_dueling_network=True, dueling_type='avg', target_model_update=1e-2, policy=policy)
+dqn.compile(Adam(lr=1e-3), metrics=['mae'])
```
Comment (the-sea) [inline examples/duel_dqn_cartpole.py]: LGTM

**Comment by raopku** [comment]
@matthiasplappert 
I can not see your remaining comments(you annotated the respective line directly). Thanks!

File: tests/integration/test_discrete.py [hunk 2]
```diff
@@ -86,3 +86,27 @@ def test_cem():
...
+    # Next, we build a very simple model.
+    model = Sequential()
+    model.add(Dense(16, input_shape=(1,)))
+    model.add(Activation('relu'))
+    model.add(Dense(nb_actions, activation='linear'))
+
+    memory = SequentialMemory(limit=1000, window_length=1)
+    policy = EpsGreedyQPolicy(eps=.1)
+    dqn = DQNAgent(model=model, nb_actions=nb_actions, memory=memory, nb_steps_warmup=50,
+                   target_model_update=1e-1, policy=policy, enable_double_dqn=False, enable_dueling_network=True)
```
Comment (raopku) [inline tests/integration/test_discrete.py]: LGTM

**Comment by matthiasplappert** [comment]
Sorry, I failed to submit the review... Should be visible now.

File: rl/agents/dqn.py [hunk 3]
```diff
@@ -100,6 +100,33 @@ def __init__(self, model, policy=None, enable_double_dqn=True,
...
+        self.enable_dueling_network = enable_dueling_network
+        self.advantage = dueling_type
+        if self.enable_dueling_network is True:
+            # get the second last layer of the model, abandon the last layer
+            layer = model.layers[-2]
+            nb_action = model.output._keras_shape[-1]
+            # layer y has a shape (nb_action+1,)
+            # y[:,0] represents V(s;theta)
+            # y[:,1:] represents A(s,a;theta)
+            y = Dense(nb_action + 1, activation='relu')(layer.output)
```
Comment (raopku) [inline rl/agents/dqn.py]: You are right!

File: rl/agents/dqn.py [hunk 4]
```diff
@@ -100,6 +100,30 @@ def __init__(self, model, policy=None, enable_double_dqn=True,
 
         # Parameters.
         self.enable_double_dqn = enable_double_dqn
+        self.enable_dueling_network = enable_dueling_network
+        self.advantage = dueling_type
+        if self.enable_dueling_network is True:
```
Comment (raopku) [inline rl/agents/dqn.py]: self.enable_dueling_network is sufficient

**Comment by matthiasplappert** [comment]
All done! Thanks again for your contribution.

**Pull request merged**